src/compile/data/optimize.ts

```diff
@@ -121,76 +127,80 @@ function getLeaves(roots: DataFlowNode[]) {
 /**
  * Inserts an Intermediate ParseNode containing all non-conflicting Parse fields and removes the empty ParseNodes
  */
-export function mergeParse(node: DataFlowNode): optimizers.OptimizerFlags {
-  let mutatedFlag = false;
-  const parent = node.parent;
-  const parseChildren = parent.children.filter((x): x is ParseNode => x instanceof ParseNode);
-
-  if (parseChildren.length > 1) {
-    const commonParse = {};
-    for (const parseNode of parseChildren) {
-      const parse = parseNode.parse;
-      for (const k of keys(parse)) {
-        if (commonParse[k] === undefined) {
-          commonParse[k] = parse[k];
-        } else if (commonParse[k] !== parse[k]) {
-          delete commonParse[k];
-        }
-      }
-    }
-    if (keys(commonParse).length !== 0) {
-      mutatedFlag = true;
-      const mergedParseNode = new ParseNode(parent, commonParse);
+export class MergeParse extends BottomUpOptimizer {
+  public optimize(node: DataFlowNode): optimizers.OptimizerFlags {
+    const parent = node.parent;
+    const parseChildren = parent.children.filter((x): x is ParseNode => x instanceof ParseNode);
+
+    if (parseChildren.length > 1) {
+      const commonParse = {};
       for (const parseNode of parseChildren) {
-        for (const key of keys(commonParse)) {
-          delete parseNode.parse[key];
+        const parse = parseNode.parse;
+        for (const k of keys(parse)) {
+          if (commonParse[k] === undefined) {
+            commonParse[k] = parse[k];
+          } else if (commonParse[k] !== parse[k]) {
+            delete commonParse[k];
+          }
         }
-        parent.removeChild(parseNode);
-        parseNode.parent = mergedParseNode;
-        if (keys(parseNode.parse).length === 0) {
-          parseNode.remove();
+      }
+      if (keys(commonParse).length !== 0) {
+        this.setMutated();
+        const mergedParseNode = new ParseNode(parent, commonParse);
+        for (const parseNode of parseChildren) {
+          for (const key of keys(commonParse)) {
+            delete parseNode.parse[key];
+          }
+          parent.removeChild(parseNode);
+          parseNode.parent = mergedParseNode;
+          if (keys(parseNode.parse).length === 0) {
+            parseNode.remove();
+          }
         }
       }
     }
+    this.setContinue();
+    return this.flags;
   }
-  return {continueFlag: true, mutatedFlag};
 }
 
 export function isTrue(x: boolean) {
   return x;
 }
 
+// any type to avoid error of Cannot use 'new' with an expression whose type lacks a call or construct signature.
```

Comment:
I don't get this comment

Reply:
Wouldn't it be better if this was a method of the class? Something like

const opt = new RemoveNodeOptimizer(...).
Opt.run()

Reply:
It's a bit uglier, though. 

Before you do this, let's talk a bit more high level. What's the advantage of using classes? I feel one is type safety and with the any you loose that, no?

Reply:
To make code more readable, the first argument should be the class!